main.nf

```diff
@@ -167,22 +245,333 @@ process fastqc {
+
+    script:
+    if ( !params.singleEnd ) {
+    """
+    bowtie2-build --threads "${task.cpus}" "${assembly}" ref
+    bowtie2 -p "${task.cpus}" -x ref -1 "${reads[0]}" -2 "${reads[1]}" | \
+        samtools view -@ "${task.cpus}" -bS | \
+        samtools sort -@ "${task.cpus}" -o "${name}.bam"
+    samtools index "${name}.bam"
+    jgi_summarize_bam_contig_depths --outputDepth depth.txt "${name}.bam"
+    metabat2 -t "${task.cpus}" -i "${assembly}" -a depth.txt -o "bins/${name}" -m ${min_size}
+    """
+    }
+    else {
+    """
+    bowtie2-build --threads "${task.cpus}" "${assembly}" ref
+    bowtie2 -p "${task.cpus}" -x ref -U ${reads} | \
+        samtools view -@ "${task.cpus}" -bS | \
+        samtools sort -@ "${task.cpus}" -o "${name}.bam"
+    samtools index "${name}.bam"
+    jgi_summarize_bam_contig_depths --outputDepth depth.txt "${name}.bam"
+    metabat2 -t "${task.cpus}" -i "${assembly}" -a depth.txt -o bins/"${name}" -m ${min_size}
+    """
+    }
+
+}
+
+
+process checkm_download_db {
+    output:
+        file("checkm_data") into checkm_db
+
+    when:
+        params.no_checkm == false
+
+    script:
+    """
+    mkdir -p checkm_data && \
+    cd checkm_data && \
+    curl -L -O https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz && \
+    tar xzf checkm_data_2015_01_16.tar.gz && \
+    cd .. && \
+    printf "checkm_data\ncheckm_data\n" | checkm data setRoot
+    """
+}
+
+process checkm {
+    tag "$name"
+    publishDir "${params.outdir}/checkm", mode: 'copy'
+
+    input:
+    set val(name), file(bins) from metabat_bins
+    set val(_name), file(bam) from mapped_reads_checkm
+    val(delta_cont) from params.delta_cont
+    val(merged_cont) from params.merged_cont
+    val(delta_compl) from params.delta_compl
+    val(abs_delta_cov) from params.abs_delta_cov
+    val(delta_gc) from params.delta_gc
+    file("checkm_data") from checkm_db
+
+    output:
+    file("${name}_stats/lineage") into checkm_merge_results
+    file("${name}_stats/plots") into checkm_merge_plots
+    file("${name}_stats/qa.txt") into checkm_merge_qa
+    set val(name), file("${name}/") into checkm_merge_bins
+
+    when:
+        params.no_checkm == false
+
+    script:
+    """
+    # re-run setRoot in case checkm has forgotten where the databases are located
+    chd=\$(readlink -f checkm_data)
+    printf "\$chd\\n\$chd\\n" | checkm data setRoot
+
+    mkdir -p stats
+    checkm lineage_wf -t "${task.cpus}" -x fa "${bins}" stats/lineage > stats/qa.txt
+    checkm bin_qa_plot -x fa stats/lineage "${bins}" stats/plots
+
+    samtools index "${bam}"
+    checkm coverage -t "${task.cpus}" -x fa "${bins}" stats/coverage.txt "${bam}"
+    checkm profile stats/coverage.txt > stats/profile.txt
+    checkm tree_qa stats/lineage > stats/tree_qa.txt
+
+    samtools view "${bam}" | awk '{if (NR <=1000) print length(\$10)}' >  stats/read_length.txt
```

Comment:
I had to look it up, but `NR` is as far as I understand it the number of rows. So, you're only printing the first 1000 rows? However, I think `awk` will evaluate all rows from `samtools view` even after the first 1000, so this is actually much faster if piped through `head -n 1000` before it goes into awk.